pkg/reconciler/pipelinerun/resources/pipelinerunresolution_test.go

```diff
@@ -741,76 +582,100 @@ func TestGetNextTaskWithRetries(t *testing.T) {
-		state: taskSucceededState,
-		candidates: map[string]struct{}{
-			"mytask5": {},
-		},
+		name:         "tasks-cancelled-no-candidates",
+		state:        taskCancelledByStatusState,
 		expectedNext: []*ResolvedPipelineRunTask{},
 	}, {
-		name:  "tasks-retried-no-candidates",
-		state: taskRetriedState,
-		candidates: map[string]struct{}{
-			"mytask5": {},
-		},
+		name:         "tasks-cancelled-bySpec-no-candidates",
+		state:        taskCancelledBySpecState,
 		expectedNext: []*ResolvedPipelineRunTask{},
 	}, {
-		name:  "tasks-retried-one-candidates",
-		state: taskExpectedState,
-		candidates: map[string]struct{}{
-			"mytask5": {},
-		},
+		name:         "tasks-running-no-candidates",
+		state:        taskRunningState,
+		expectedNext: []*ResolvedPipelineRunTask{},
+	}, {
+		name:         "tasks-succeeded-bySpec-no-candidates",
+		state:        taskSucceededState,
+		expectedNext: []*ResolvedPipelineRunTask{},
+	}, {
+		name:         "tasks-retried-no-candidates",
+		state:        taskRetriedState,
+		expectedNext: []*ResolvedPipelineRunTask{},
+	}, {
+		name:         "tasks-retried-one-candidates",
+		state:        taskExpectedState,
 		expectedNext: []*ResolvedPipelineRunTask{taskExpectedState[0]},
 	}}
-
-	// iterate over *state* to get from candidate and check if TaskRun is there.
-	// Cancelled TaskRun should have a TaskRun cancelled and with a retry but should not retry.
-
 	for _, tc := range tcs {
 		t.Run(tc.name, func(t *testing.T) {
-			next := tc.state.GetNextTasks(tc.candidates)
+			tcDag, err := DagFromState(tc.state)
+			if err != nil {
+				t.Fatalf("Something went wrong compiling the TC: %v", err)
+			}
+			rpr := ResolvedPipelineRun{
+				Dag:      tcDag,
+				DagState: tc.state,
+				WontRun:  skippedTasks(tc.state, tcDag),
+			}
+			next, err := rpr.GetNextTasks()
+			if err != nil {
+				t.Fatalf("Somehow failed to get next tasks: %v", err)
+			}
 			if d := cmp.Diff(next, tc.expectedNext); d != "" {
 				t.Errorf("Didn't get expected next Tasks %s", diff.PrintWantGot(d))
 			}
 		})
 	}
 }
 
-func TestIsDone(t *testing.T) {
+func TestGetNextTasksError(t *testing.T) {
+	// TBD One test with invalid state (ancestor not done, child done)
+}
+
+func TestIsDAGDone(t *testing.T) {
 
+	// This task is marked as failed because it was cancelled.
+	// On a normal failure it would retry, but it shouldn't here because it was
+	// cancelled.
 	var taskCancelledByStatusState = PipelineRunState{{
 		PipelineTask: &pts[4], // 2 retries needed
 		TaskRunName:  "pipelinerun-mytask1",
-		TaskRun:      withCancelled(makeRetried(trs[0])),
+		TaskRun:      withCancelled(makeFailed(trs[0])),
 		ResolvedTaskResources: &resources.ResolvedTaskResources{
 			TaskSpec: &task.Spec,
 		},
 	}}
 
+	// The cancelled by spec here has no impact on how this is treated by any
+	// of the functions in pipelineresolution. The `makeRetried` is what triggers
+	// the expected behaviour.
+	// TODO: remove withCancelledBySpec tests as misleading
```

Comment:
I find this test misleading and it's not really testing anything to do with `withCancelledBySpec`.
It does test that a failure when retries are exhausted marks the task and DAG as done, so perhaps we could remove the `withCancelledBySpec` and rename the test?